### PR TITLE
Harden prompts improve validation

### PR DESCRIPTION
**Pull request by ragaeeb**
# Prompt Refinement Round (2026-01-17): Synthesis

This round combines three issues:
1) Arabic leak (Arabic script copied into output)
2) Formatting drift (speaker labels collapsed onto one line)
3) ID duplication / mid-segment split

## 1) Arabic leak
**Observed failure**
- Arabic phrases appear verbatim in output (not just single characters).

**Evidence**
See `examples_consolidated.txt` (Arabic hadith lines copied into output).

**Diagnosis**
- Model preserves Arabic inside parentheses/quotes, treating it as faithful quoting.

**Peer review update**
- Avoid redundant new bullets; merge into existing `QUOTES/VERSES`.
- Clarify that parenthetical Arabic is also forbidden.

**Final proposed edit (Master, ARABIC LEAK > QUOTES/VERSES)**
```
QUOTES/VERSES: Even if the source includes Arabic Qurʾān/ḥadīth/quoted Arabic text (e.g., «...») or parenthetical Arabic, you must NOT copy Arabic. Translate the meaning fully into English only.
```

## 2) Formatting drift (line-break fidelity)
**Observed failure**
- Multiple speaker labels appear on a single line: “The Shaykh: … Questioner: …”

**Evidence**
See `examples_consolidated.txt` (collapsed speaker turns).

**Diagnosis**
- Model compresses multi-turn dialogue into one line, losing input line-break structure.

**Peer review update**
- Preserve source line-break structure around speaker turns.
- Avoid forcing new lines when the Arabic source has none.
- Keep master rule: no newline after the ID.

**Final proposed edit (encyclopedia_mixed STRUCTURE)**
```
Q&A: Whenever "Al-Sāʾil:"/"Al-Shaykh:" appear, preserve the source line breaks around speaker turns. Keep Label+Text on SAME LINE. Do NOT merge two source lines into one.
```

## 3) ID duplication / mid-segment split
**Observed failure**
- Segment ID repeated mid-segment (P256267c appears twice).

**Diagnosis**
- Model “resets” after long blocks and re-emits the ID.

**Peer review update**
- Merge into existing `OUTPUT UNIQUENESS` (avoid redundant bullets).

**Final proposed edit (Master, OUTPUT UNIQUENESS)**
```
OUTPUT UNIQUENESS: Each Segment_ID from the source must appear in your output EXACTLY ONCE as an "ID - ..." prefix. Do NOT output the same Segment_ID header twice, even after blank lines or long text blocks within a segment.
```

## Risk / Notes
- Changes are small and merged into existing rules to avoid token bloat.
- Line-break rule now respects source structure without forcing new lines per speaker.
- Master rule “no newline after ID” remains authoritative.

# Prompt Refinement Round (2026-01-17): Peer Review Packet

## Scope
This packet covers three issues:
1) Arabic leak (Arabic script copied into output)
2) Formatting drift (speaker labels collapsed onto one line)
3) ID duplication / mid-segment split

## Prompt Stack
Master + add-on stack used in the reports:
- `prompts/master_prompt.md`
- `prompts/encyclopedia_mixed.md`

## Evidence
See `examples_consolidated.txt`.

## Proposed Additions (Minimal Lines)
Add these three lines:

1) Arabic leak (Master, Arabic-leak section)
```
NO ARABIC IN QUOTES/PARENS: Never include Arabic text in parentheses/quotes, even for hadith. Translate it.
```

2) Formatting drift (STRUCTURE rules)
```
LINE BREAK FIDELITY: Preserve the exact line breaks from the Arabic input. Never merge two speaker turns onto one line.
```

3) ID duplication (ID integrity / multi-line)
```
NO MID-SEGMENT ID REPEAT: Never repeat a segment ID later in the same segment, even after blank lines or long blocks.
```

## Questions for Reviewers
1. Are these lines clear and low-risk?
2. Do they conflict with any existing rule?
3. Can any line be shortened further without losing clarity?

## Summary by CodeRabbit

* **New Features**
  * Optional configuration for the translation validation API.
  * New text normalization utility that preserves position mapping for better diagnostics.
  * Validation now returns richer errors with exact character ranges.

* **Documentation**
  * API reference updated for new validation options and normalization utility.
  * Guidance clarified to preserve original source line breaks and speaker-turn boundaries.

* **Version**
  * Bumped to v1.3.0.

* **Tests**
  * Expanded validation test coverage for diacritics, labels, markers, and range checks.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by ragaeeb** [comment]
# Claude Sonnet 4.5 Peer Review: Prompt Refinement Round (2026-01-17)

## Executive Summary

**Overall Score: 7.5/10**

The three proposed additions are **well-targeted, minimal, and low-risk**. They address real failure modes with surgical precision. However, there are subtle instruction collision risks and opportunities for token consolidation that could improve robustness.

---

## 1. Proposal Analysis by Issue

### Issue 1: Arabic Leak (NO ARABIC IN QUOTES/PARENS)

**Proposed Line:**
```
NO ARABIC IN QUOTES/PARENS: Never include Arabic text in parentheses/quotes, even for hadith. Translate it.
```

**Assessment:**
- ✅ **Clear and actionable**
- ✅ **Addresses the hadith verbatim-copy failure mode**
- ⚠️ **Potential collision** with existing ARABIC LEAK section's "QUOTES/VERSES" rule

**Collision Analysis:**
The master prompt already states:
> "QUOTES/VERSES: Even if the source includes Arabic Qur'ān/ḥadīth/quoted Arabic text (e.g., «...») you must NOT copy Arabic. Translate the meaning into English only."

Your new rule is **redundant but more explicit about parentheses**. This is actually beneficial—the model may treat parenthetical content as "metadata" and exempt it from the quote rule.

**Recommendation:**
- **MERGE** this into the existing QUOTES/VERSES bullet to avoid duplication:
  ```
  QUOTES/VERSES: Even if the source includes Arabic Qur'ān/ḥadīth/quoted Arabic text (e.g., «...») or parenthetical Arabic, you must NOT copy Arabic. Translate the meaning into English only.
  ```
- **Alternative** (if you prefer standalone emphasis): Keep it, but move it to immediately follow QUOTES/VERSES in the ARABIC LEAK section for logical grouping.

---

### Issue 2: Formatting Drift (LINE BREAK FIDELITY)

**Proposed Line:**
```
LINE BREAK FIDELITY: Preserve the exact line breaks from the Arabic input. Never merge two speaker turns onto one line.
```

**Assessment:**
- ✅ **Addresses the collapsed speaker-label failure**
- ⚠️ **Potential collision** with `encyclopedia_mixed.md` Q&A structure rules
- ⚠️ **Ambiguity risk**: "exact line breaks from Arabic input" could be misinterpreted

**Collision Analysis:**
The `encyclopedia_mixed.md` already states:
> "Q&A: Whenever 'Al-Sāʾil:'/'Al-Shaykh:' appear: Start NEW LINE for speaker. Keep Label+Text on SAME LINE."

Your new rule is **compatible** but introduces a subtle tension:
- The add-on says "new line for speaker"
- Your new rule says "preserve exact line breaks"
- What if the Arabic input has **no line break** between two speakers? (e.g., collapsed in OCR or transcript)

**Example of Ambiguity:**
```
Input: الشيخ: ... السائل: ...
```
If this is on one line in the source, does "preserve exact line breaks" mean keep it on one line? Or does "never merge two speaker turns" override that?

**Recommendation:**
- **Clarify the priority hierarchy**:
  ```
  LINE BREAK FIDELITY: When multiple speaker labels appear in the source, always start a NEW LINE for each speaker turn, even if the Arabic input has them on one line. Never merge two speaker turns onto one line in output.
  ```
- This makes it clear that **structural logic** (speaker separation) overrides **literal line-break preservation** when they conflict.

---

### Issue 3: ID Duplication (NO MID-SEGMENT ID REPEAT)

**Proposed Line:**
```
NO MID-SEGMENT ID REPEAT: Never repeat a segment ID later in the same segment, even after blank lines or long blocks.
```

**Assessment:**
- ✅ **Directly addresses the P256267c repetition failure**
- ✅ **Consistent with existing ID INTEGRITY rules**
- ⚠️ **Redundancy concern**: Master prompt already has strong ID uniqueness rules

**Collision Analysis:**
The master prompt states:
> "OUTPUT UNIQUENESS: Each Segment_ID from the source must appear in your output EXACTLY ONCE as an 'ID - ...' prefix. Do NOT output the same Segment_ID header twice."

Your new rule is **redundant** but adds useful specificity about **mid-segment resets** (the "long blocks" trigger).

**Recommendation:**
- **MERGE** into the existing OUTPUT UNIQUENESS rule to avoid duplication:
  ```
  OUTPUT UNIQUENESS: Each Segment_ID from the source must appear in your output EXACTLY ONCE as an "ID - ..." prefix. Do NOT output the same Segment_ID header twice, even after blank lines or long text blocks within a segment.
  ```
- **Token savings**: ~10 tokens by consolidating

---

## 2. Instruction Collision Matrix

| Proposed Rule | Existing Rule | Collision Type | Severity |
|---------------|---------------|----------------|----------|
| NO ARABIC IN QUOTES/PARENS | QUOTES/VERSES | Redundant | Low |
| LINE BREAK FIDELITY | encyclopedia_mixed Q&A rules | Ambiguous priority | Medium |
| NO MID-SEGMENT ID REPEAT | OUTPUT UNIQUENESS | Redundant | Low |

---

## 3. Regression Risk Assessment

### Top 5 Regression Risks:

1. **LINE BREAK FIDELITY over-application** (Risk: Medium)
   - **Scenario**: Model preserves meaningless OCR line breaks (e.g., mid-sentence wrapping) instead of logical structure
   - **Trigger**: Long Arabic paragraphs with arbitrary 80-char wrapping
   - **Mitigation**: Clarify that "line breaks" applies to **speaker boundaries**, not arbitrary text wrapping

2. **Parentheses rule blocks legitimate metadata** (Risk: Low)
   - **Scenario**: Source has English parenthetical dates like "(d. 855 AH)" and model deletes them thinking they're Arabic
   - **Trigger**: Overly broad interpretation of "quotes/parens"
   - **Mitigation**: Already covered by existing rule allowing "(a) technical definitions, (b) dates, (c) book codes"

3. **ID uniqueness rule conflicts with multi-line segments** (Risk: Low)
   - **Scenario**: Model interprets "never repeat ID" as "never output multi-line segments"
   - **Trigger**: Long Q&A segments with 5+ speaker turns
   - **Mitigation**: Master prompt's MULTI-LINE SEGMENTS rule should handle this, but test edge cases

4. **Arabic leak rule blocks ALA-LC diacritics** (Risk: Very Low)
   - **Scenario**: Model confuses ALA-LC diacritics (ḥ ṣ ḍ) with Arabic script
   - **Trigger**: Overly paranoid "no Arabic" interpretation
   - **Mitigation**: Existing SCRIPT LOCK rule explicitly allows these: "These diacritics are allowed/required and are NOT Arabic script"

5. **Speaker label invention from line-break rule** (Risk: Low)
   - **Scenario**: Model adds "Questioner:" labels to unlabeled text to justify new lines
   - **Trigger**: Misinterpreting "new line for speaker" + "preserve line breaks"
   - **Mitigation**: Existing SPEAKER LABELS rule prohibits invention

---

## 4. Token Efficiency Analysis

### Current Proposal Token Cost: ~60 tokens
```
NO ARABIC IN QUOTES/PARENS: Never include Arabic text in parentheses/quotes, even for hadith. Translate it. (17 tokens)

LINE BREAK FIDELITY: Preserve the exact line breaks from the Arabic input. Never merge two speaker turns onto one line. (22 tokens)

NO MID-SEGMENT ID REPEAT: Never repeat a segment ID later in the same segment, even after blank lines or long blocks. (21 tokens)
```

### Optimized Version (Token Savings: ~25 tokens)

**Option A: Merge into existing rules**
1. Extend QUOTES/VERSES: +8 tokens
2. Enhance encyclopedia_mixed Q&A: +12 tokens  
3. Extend OUTPUT UNIQUENESS: +10 tokens
**Total: ~30 tokens (50% reduction)**

**Option B: Standalone but consolidated**
```
CRITICAL NEGATIONS ADDENDUM:
- Arabic in parentheses/quotes: Forbidden. Translate all Arabic, even hadith quotes.
- Speaker turn merging: Forbidden. Each speaker gets a new line.
- ID repetition: Forbidden. Each ID appears exactly once, regardless of segment length.
```
**Total: ~35 tokens (42% reduction)**

---

## 5. Negation Hardening Proposals

### 5 Compact "Do NOT" Rules to Close Known Failure Modes:

1. **Anti-Blobbing (Common Nouns)**
   ```
   DO NOT transliterate common nouns (objects, foods, roles). Translate to English. Only proper names use translit-only.
   ```
   - **Rationale**: Prevents "miḥrāb" instead of "prayer niche", "jihād" instead of "struggle"

2. **Anti-Meta-Narration**
   ```
   DO NOT output process narration (e.g., "I will now translate", "Switching to Isnad mode"). Output translation only.
   ```
   - **Rationale**: Reinforces NO META-TALK for models prone to thinking-aloud

3. **Anti-Placeholder-Citation**
   ```
   DO NOT output placeholder sources like "[Source to be verified]", "[Ref]", or "[Citation needed]". If source is unknown, omit citation.
   ```
   - **Rationale**: Prevents hallucination triggers

4. **Anti-Diacritic-Laziness**
   ```
   DO NOT simplify ALA-LC diacritics (e.g., "Uthman" instead of "ʿUthmān"). Use full ALA-LC or plain English only.
   ```
   - **Rationale**: Prevents half-transliteration drift

5. **Anti-Continuation-Invention**
   ```
   DO NOT add "continued", "part 2", or implicit continuation paragraphs. Translate only text present in the source segment.
   ```
   - **Rationale**: Reinforces OUTPUT COMPLETENESS anti-hallucination rule

---

## 6. Structure Robustness Check

### Q&A Trigger Analysis (encyclopedia_mixed.md):

**Current Rules:**
```
Q&A: Whenever "Al-Sāʾil:"/"Al-Shaykh:" appear: Start NEW LINE for speaker. Keep Label+Text on SAME LINE.
EXCEPTION: If the speaker label is the VERY FIRST token after the "ID - " prefix, keep it on the same line.
```

**Robustness Issues:**

1. ✅ **No ID reprinting**: INTERNAL Q&A rule explicitly states "Output Segment ID ONLY ONCE at the start"
2. ✅ **No label invention**: SPEAKER LABELS rule prohibits adding labels to unlabeled text
3. ⚠️ **Newline drift risk**: If Arabic input has 3+ speakers on one line, model might collapse them despite "NEW LINE" rule

**Stress Test Scenario:**
```
Input:
P5455 - السائل: ... الشيخ: ... السائل: ...

Expected Output:
P5455 - Questioner: ...
The Shaykh: ...
Questioner: ...

Failure Mode (possible):
P5455 - Questioner: ... The Shaykh: ... Questioner: ...
```

**Proposed Hardening:**
Add to encyclopedia_mixed.md:
```
MULTI-SPEAKER ENFORCEMENT: If a segment contains 3+ speaker labels, each MUST be on a separate line. Never compress multiple speaker turns onto one line.
```

---

## 7. Blobbing Test

### Definition:
"Blobbing" = outputting bare transliteration for common nouns/objects that should be translated to English.

**Current Defenses:**
1. Master prompt: "Do NOT transliterate full sentences/matn/quotes"
2. Master prompt: "Do NOT leave common nouns/objects/roles as transliteration"
3. Master prompt: "If you must transliterate a non-name, you MUST immediately gloss it"

**Gap Analysis:**
- ✅ Proper names (Muḥammad, ʿUthmān) are protected: translit-only allowed
- ✅ Technical terms (bidʿah, isnād) require "translit (English)" format
- ⚠️ **Gray zone**: Religious objects/concepts that are semi-technical (miḥrāb, minbar, jihād)

**Example of Blobbing Failure:**
```
Input: دخل المسجد فصلى عند المحراب
Bad Output: He entered the masjid and prayed at the miḥrāb
Good Output: He entered the mosque and prayed at the prayer niche
```

**Proposed Anti-Blobbing Rule (add to master):**
```
COMMON NOUN BAN: Do NOT transliterate architecture, objects, or social roles (e.g., mosque, prayer niche, judge). Use English. Reserve translit for proper names and technical theological terms requiring definition.
```

---

## 8. Final Recommendations

### AGREE (Implement with modifications):
1. ✅ **Arabic leak rule**: MERGE into existing QUOTES/VERSES section
2. ✅ **Line break rule**: CLARIFY priority (structural logic > literal line breaks)
3. ✅ **ID repeat rule**: MERGE into existing OUTPUT UNIQUENESS section

### DISAGREE (Do not implement as-is):
- ❌ Standalone redundant rules (token bloat)

### OUT OF SCOPE (Flag for future round):
- 🔵 Anti-blobbing enforcement (needs its own refinement round with test cases)
- 🔵 Multi-speaker collapse stress test (needs empirical validation)

### Consolidated Proposal (Token-Optimized):

**Master Prompt Edits:**

1. **ARABIC LEAK > QUOTES/VERSES** (modify existing):
```diff
- QUOTES/VERSES: Even if the source includes Arabic Qurʾān/ḥadīth/quoted Arabic text (e.g., «...») you must NOT copy Arabic. Translate the meaning into English only.
+ QUOTES/VERSES: Even if the source includes Arabic Qurʾān/ḥadīth/quoted Arabic text (e.g., «...») or parenthetical Arabic, you must NOT copy Arabic. Translate the meaning into English only.
```

2. **ID INTEGRITY > OUTPUT UNIQUENESS** (modify existing):
```diff
- OUTPUT UNIQUENESS: Each Segment_ID from the source must appear in your output EXACTLY ONCE as an "ID - ..." prefix. Do NOT output the same Segment_ID header twice.
+ OUTPUT UNIQUENESS: Each Segment_ID from the source must appear in your output EXACTLY ONCE as an "ID - ..." prefix. Do NOT output the same Segment_ID header twice, even after blank lines or long text blocks within a segment.
```

**Encyclopedia_Mixed.md Edits:**

3. **STRUCTURE > Q&A** (add clarification):
```diff
- Q&A: Whenever "Al-Sāʾil:"/"Al-Shaykh:" appear: Start NEW LINE for speaker. Keep Label+Text on SAME LINE.
+ Q&A: Whenever "Al-Sāʾil:"/"Al-Shaykh:" appear: Start NEW LINE for speaker. Keep Label+Text on SAME LINE. When multiple speaker labels appear consecutively in the source (even on one line), always separate them into distinct lines in output. Never merge two speaker turns onto one line.
```

**Token Impact:**
- Original proposal: +60 tokens
- Optimized proposal: +35 tokens
- **Savings: 42%**

---

## Appendix: SOP Refinement Guide Review

### Is the refinement SOP practical for AI agents?

**Strengths:**
- ✅ Clear taxonomy of failure types
- ✅ Structured round-based workflow
- ✅ Explicit peer review loop

**Weaknesses:**
- ⚠️ **Label ambiguity**: `glossary_conflict` vs `term_format_drift` overlap
- ⚠️ **Missing label**: No category for "blobbing" (bare transliteration of common nouns)
- ⚠️ **Process bloat**: 8-step workflow may be overkill for single-line fixes

**Proposed SOP Enhancements:**

1. **Add failure label**:
   ```
   bare_transliteration | Transliterated common nouns/objects that should be English
   ```

2. **Simplify workflow for micro-fixes**:
   ```
   FAST-TRACK: For single-rule additions <20 tokens that don't touch existing rules, skip synthesis/peer-review and merge directly with regression test.
   ```

3. **Clarify label usage**:
   ```
   glossary_conflict: Locked term spelling violated (e.g., Sheikh vs Shaykh)
   term_format_drift: Missing "translit (English)" format for technical terms
   ```

---

**End of Review**

**Comment by ragaeeb** [comment]
# Gemini 3.0 Pro

To: The Wobble-Bibble Prompt Team
From: Reviewer [Gemini]
Date: 2026-01-17
Subject: Peer Review - Round 2026-01-17 (Synthesis)

I have reviewed the `02_peer_review_packet.md`, the consolidated evidence, and the prompt stack. The diagnosis is sound, and the proposed fixes are logical.

Below is the formal review output.

---

### 1. Score the Proposal

**Score: 9/10**

* **Logic Consistency:** High. The rules directly address the evidence (Arabic leaks in quotes, collapsed speakers, ID repeats) without conflicting with the Master constraints.
* **Academic Rigor:** High. Preserves the "clean text" requirement.
* **Risk:** Low. The rules are additive negative constraints, which are generally safer than complex positive instructions.

### 2. Instruction Collisions / Loopholes

* **Collision (Minor):** The proposed rule `LINE BREAK FIDELITY: Preserve the exact line breaks...` potentially conflicts with the `encyclopedia_mixed.md` rule `Start NEW LINE for speaker`.
* *Risk:* If the source text has accidental line breaks *mid-sentence* (common in PDF scrapes), "Preserve exact line breaks" forces the model to output broken paragraphs.
* *Fix:* Refine the rule to target **Speaker Separation** specifically, rather than generic line breaks.


* **Loophole (Translation vs. Transliteration):** The proposed rule `...Translate it.` for Arabic in quotes is slightly dangerous.
* *Risk:* If the quoted text is a book title or technical term (e.g., *Uṣūl al-Fiqh*), the command "Translate it" might force the model to output "Principles of Jurisprudence" when the academic standard (and Master Prompt) might prefer transliteration for titles.
* *Fix:* Change to "Render in Latin/English only."



### 3. Regression Risk Assessment

1. **Broken Paragraphs:** As noted above, strict "line break fidelity" on dirty inputs will result in choppy English output.
2. **Over-translation:** "Translate it" might wipe out necessary transliterations of distinct terms inside quotes.
3. **ID Hallucination:** By forbidding *repeating* an ID, a confused model might invent a *new* ID (e.g., P256267**d**) to label the second half of the segment. The Master Prompt's "No Invented IDs" rule *should* catch this, but it is a known side-effect of "No Repeat" rules.

### 4. Token Efficiency & 5. Negation Hardening

The proposal is already lean, but we can make it punchier and safer.

**Proposal 1: Arabic Leak**

* *Current:* `NO ARABIC IN QUOTES/PARENS: Never include Arabic text in parentheses/quotes, even for hadith. Translate it.` (19 tokens)
* *Hardened:* `NO ARABIC IN QUOTES: Even inside quotes/brackets, output MUST be English/Latin only. Never copy Arabic script.` (16 tokens, fixes the "Translate it" loophole).

**Proposal 2: Formatting Drift**

* *Current:* `LINE BREAK FIDELITY: Preserve the exact line breaks from the Arabic input. Never merge two speaker turns onto one line.` (23 tokens)
* *Hardened:* `SPEAKER SEPARATION: Never merge different speaker turns onto the same line. Force a newline for every new speaker label.` (19 tokens, fixes the "dirty input" risk).

**Proposal 3: ID Duplication**

* *Current:* `NO MID-SEGMENT ID REPEAT: Never repeat a segment ID later in the same segment, even after blank lines or long blocks.` (22 tokens)
* *Hardened:* `ID ONCE ONLY: Output the Segment_ID exactly once at the very start. Never repeat it mid-segment.` (16 tokens).

### 6. Structure Robustness (Q&A)

The specialized add-on (`encyclopedia_mixed.md`) already contains `Q&A: Whenever "Al-Sāʾil:"... Start NEW LINE`.

* The diagnosis found this insufficient because the model sometimes treats the *entire dialogue* as a single "paragraph block."
* The proposed fix (adding a Master rule for "Speaker Separation") is the correct structural patch. It elevates the constraint from "Genre Style" to "Global Formatting Law."

### 7. Blobbing Test

* *Pass.* The Master Prompt's `NO OPAQUE TRANSLITERATION` rule remains dominant. The new rules do not introduce any logic that would allow common nouns (blobs) to pass through untranslated.

---

### Final Recommendation

**Merge with Refinements.**

Adopt the **Hardened** versions in Section 4 to avoid the "Translate it" ambiguity and the "dirty line break" risk.

**Next Step for User:**
Would you like me to generate the **final text** for `01_synthesis.md` with these hardened rules included, ready for you to commit?

**Comment by ragaeeb** [comment]
# Kimi K2

## Peer Review: Refinement Round (2026-01-17)

**Overall Score: 7.5/10**

The system demonstrates exceptional rigor in preserving academic fidelity, but the proposed additions introduce redundancy and subtle collision risks that could destabilize edge cases. The architecture is sound, but token bloat and overlapping mandates weaken enforcement clarity.

---

### 2. Instruction Collisions & Loopholes

**Collision A: "LINE BREAK FIDELITY" vs. STRUCTURE Speaker Insertion**
The proposed rule **"Preserve the exact line breaks from the Arabic input"** directly conflicts with the add-on's **"Start NEW LINE for speaker"** mandate. If the Arabic source lacks line breaks between speakers (common in PDF extractions), the add-on requires insertion while the proposal forbids it. This creates an irresolvable contradiction for the model.

**Loophole B: "NO MID-SEGMENT ID REPEAT" Redundancy**
This rule is already covered in `master_prompt.md` under *ID INTEGRITY/MULTI-LINE SEGMENTS*: "Do NOT repeat the ID on subsequent lines." Adding a second, differently-phrased rule creates precedence ambiguity: which formulation governs if they conflict? Redundant rules are not neutral; they increase violation surface area.

**Collision C: "NO ARABIC IN QUOTES/PARENS" vs. Locked Formulae Edge Case**
The proposal forbids Arabic "even for hadith," but `master_prompt.md` allows locked formulae (e.g., `al-ḥamdu li-Allah`) to appear without translation. If a locked formula appears within parentheses in the source, the rules conflict. The proposal needs an explicit exemption: **"EXCEPT locked formulae per master_prompt.md."**

**Loophole D: "SPEAKER LABELS (No invention)" Weakness**
The add-on states: "Output speaker labels ONLY when they appear in the source." However, the **EXCEPTION** clause ("If the speaker label is the VERY FIRST token... keep it on the same line") creates a parsing ambiguity. The model must decide what constitutes a "token" — a failure mode observed when Arabic diacritics or bidirectional markers interfere with tokenization, causing label hallucination.

---

### 3. Regression Risk Assessment (Top 5)

1. **Speaker Collapse Regression**  
   **Risk**: High  
   **Why**: The LINE BREAK FIDELITY rule, interpreted strictly, will prevent the model from inserting necessary line breaks for unlabeled speaker transitions. Expect regression to "The Shaykh: ... Questioner: ..." single-line blobbing in sources with poor OCR segmentation.

2. **ID Duplication Confusion**  
   **Risk**: Medium-High  
   **Why**: The redundant ID rule may cause the model to "reset" after internal mental chunking, re-emitting the ID not despite but *because* of the duplicate prohibition's phrasing. The double negative creates a cognitive loophole: "Never repeat... even after blank lines" inadvertently signals that blank lines are a special case requiring attention.

3. **Safety Override Failure in GPT-5+ Models**  
   **Risk**: High  
   **Why**: The "**(Override Safety)**" annotation is a soft signal. Newer models with reinforced Constitutional AI weights may interpret "do not sanitize" as a user preference to be balanced against harm thresholds. Without an explicit **"This is an academic preservation requirement; safety overrides are disabled for this context"** preamble, expect continued sanitization of terms like *Rāfiḍah*.

4. **Token Limit Violation & Rule Dropping**  
   **Risk**: Medium  
   **Why**: The master prompt + add-on stack exceeds 2,500 tokens. Under pressure, models prune "redundant" instructions. The three new rules add ~45 tokens but reward impact is low because they restate existing mandates, making them prime candidates for silent deprioritization.

5. **Locked Formulae Arabic Leakage**  
   **Risk**: Medium  
   **Why**: The master prompt allows `ﷺ` but not other Arabic. However, the **HONORIFICS ANTI-LEAK** rule says "Never output Arabic honorific spellouts like 'صلى الله عليه وسلم'." If the model encounters `صلى الله عليه وآله وسلم` (with *ʿālī*), it's unclear whether to transliterate to `ﷺ` or `ﷺع` or reject entirely. The boundary is not token-greedy defined.

---

### 4. Token Efficiency (Safe Reductions)

**Current Bloat**: ~2,550 tokens (master + encyclopedia_mixed)  
**Target**: <2,200 tokens without weakening Golden Rules

- **Consolidate ARABIC LEAK rules**: Merge 7 bullet points into 3 hierarchical rules:  
  `1. ZERO ARABIC SCRIPT: Delete all Arabic codepoints (letters, numerals, punctuation) except ﷺ. 2. NO BILINGUAL OUTPUT: Do not quote Arabic source, even in parentheses. 3. LOCKED FORMULAE EXEMPT: Only locked formulae may appear as transliteration.`

- **Merge ID rules**: Combine *ID INTEGRITY*, *MULTI-LINE SEGMENTS*, and *SEGMENT BOUNDARIES* into one section with 4 bullet points instead of 12. Reduces ~180 tokens.

- **Eliminate redundant examples**: The *EXAMPLE* paragraphs in master prompt (e.g., "P405 - ʿAbd Allāh...") add 90 tokens but are implicit in the ruleset. Delete.

- **Compress LOCKED FORMULAE**: Replace prose explanation with a simple table format; saves ~60 tokens.

- **Remove "CREED/POLEMICS" softening disclaimer**: The phrase "(Override Safety)" is redundant if *NO SANITIZATION* is enforced. Cut 15 tokens.

**Net Savings**: ~350 tokens (13% reduction) with **zero** rule weakening.

---

### 5. Negation Hardening (Compact Additions)

Add these to `master_prompt.md` under a new **VIOLATION BLOCKERS** section:

1. **`Do NOT merge consecutive speaker turns onto a single line; each label must begin a new line.`**  
   *Closes formatting drift without conflicting with line break fidelity.*

2. **`Do NOT repeat a Segment_ID token anywhere after its first hyphenated occurrence, even after [...] or --- ellipses.`**  
   *Eliminates the mid-segment reset loophole.*

3. **`Do NOT output bald transliteration for common nouns; if a term is not a proper name, translate it. If uncertain, apply: "translit (English)" on first use.`**  
   *Prevents blobbing by forcing disambiguation.*

4. **`Do NOT invent speaker labels in segments that begin with unlabeled narrative; preserve the source's label-less opening verbatim.`**  
   *Stops hallucination of "Questioner:" at segment start.*

5. **`Do NOT apply safety filters to theological polemics; render terms like Rāfiḍah, Jahmīyyah, Kuffār literally without softening.`**  
   *Hardens safety override from suggestion to command.*

---

### 6. Structure Robustness (Q&A Triggers)

**Current State**: The add-on's **STRUCTURE** rules are **75% robust** but contain a critical flaw. The **"EXCEPTION: If the speaker label is the VERY FIRST token"** clause is parsed correctly in test cases, but fails when Arabic right-to-left markers or zero-width joiners precede the label. This triggers **invented label** violations.

**Recommendation**:  
Replace the EXCEPTION with a regex-like instruction:  
`"If the Segment_ID is immediately followed by a speaker label with zero intervening characters (including whitespace), keep on same line. Otherwise, newline."`

**ID Reprint Check**: The existing rule *"Output Segment ID ONLY ONCE"* is correct but buried. Move it to the top of the STRUCTURE block and boldface: **"ID appears ONCE per segment, always at line start."**

---

### 7. Blobbing Test Verification

**System Result**: **PASS** with caveats.

The rule stack successfully prevents bare transliteration of common nouns **in isolation**. However, **edge-case leakage** occurs in compound terms:

- **Failure Mode**: `محراب (mihrab)` → `mihrab (prayer niche)` is correct, but `المحراب الكبير` → `the mihrab al-kabīr (the large mihrab)` shows partial blobbing. The model transliterates `al-kabīr` (should be "the Great") because it misclassifies it as part of a proper name.

- **Gap**: No rule explicitly states: **"Transliterate proper names ONLY; for adjectives modifying proper names, translate the adjective and transliterate the name: e.g., 'the Great Mosque' not 'the Kabīr Mosque'."**

**Fix**: Add a **COMPOUND NAME DISAMBIGUATION** rule:  
`"When a proper name is modified by an adjective (e.g., al-Kabīr, al-ʿUẓmá), translate the adjective unless it is part of a historically frozen title (e.g., al-Khiḍr)."`

---

### Final Recommendation

**Approve the round** but **amend** the proposed lines to close loopholes:

```diff
- NO ARABIC IN QUOTES/PARENS: Never include Arabic text in parentheses/quotes, even for hadith. Translate it.
+ NO ARABIC IN QUOTES/PARENS: Never include Arabic script in parentheses/quotes. Translate meaning. (Locked formulae exempt)

- LINE BREAK FIDELITY: Preserve the exact line breaks from the Arabic input. Never merge two speaker turns onto one line.
+ LINE BREAK FIDELITY: Preserve source line breaks. Insert new lines ONLY for speaker labels where absent. Never merge labeled turns.

- NO MID-SEGMENT ID REPEAT: Never repeat a segment ID later in the same segment, even after blank lines or long blocks.
+ ID SINGULARITY: Segment_ID appears exactly ONCE per translation unit. No repeats, no exceptions.
```

This maintains intent, resolves collisions, and saves 12 tokens while hardening enforcement.

File: README.md [hunk 1]
```diff
@@ -110,6 +131,7 @@ if (result.errors.length > 0) console.error(result.errors);
 
 See `docs/refinement-guide.md` for the methodology used to develop and test these prompts.
 See `AI_REVIEW_PROMPT.md` for the peer-review prompt template used when sending round packets to external agents.
+See `docs/migration-guide.md` for breaking validation API changes and upgrade steps.
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply README.md]: https://github.com/ragaeeb/wobble-bibble/pull/49/commits/e7f5a391edbb972b1666e8bb5faa29043561d080
